addons/crm/models/crm_lead.py

```diff
@@ -1468,18 +1468,38 @@ def _cron_update_automated_probabilities(self):
             leads_to_update = self.env['crm.lead'].search(pending_lead_domain)
             lead_probabilities = leads_to_update._pls_get_naive_bayes_probabilities(batch_mode=True)
 
-            # Update in execute batch to avoid server roundtrips + page_size to 10000 to avoid memory errors
+            # Group by new probability to reduce server roundtrips when executing the update
+            lead_ids_by_probabilities = defaultdict(list)
+            for lead_id, proba in sorted(lead_probabilities.items()):
+                lead_ids_by_probabilities[proba].append(lead_id)
+
             # Update both probability and automated_probability if they were equal, else, update only automated_probability
             sql = """UPDATE crm_lead
                         SET automated_probability = %s,
-                            probability = CASE WHEN (ROUND(probability::numeric, 2) = ROUND(automated_probability::numeric, 2) or probability is null)
+                            probability = CASE WHEN (probability = automated_probability or probability is null)
                                                THEN (%s)
                                                ELSE (probability)
                                                END
-                        WHERE id = %s"""
-            batch_params = [(lead_probabilities[lead.id], lead_probabilities[lead.id], lead.id) for lead in leads_to_update if lead.id in lead_probabilities]
-            extras.execute_batch(self._cr, sql, batch_params, page_size=10000)
-            _logger.info("Predictive Lead Scoring : all automated probability updated (count: %d)" % (len(leads_to_update)))
+                        WHERE id in %s"""
+
+            # Update by a limited number of leads at the same time :
+            # - avoid memory errors
+            # - avoid to block the table for too long with a too big transaction
+            step = 5
```

Comment:
Is there a basis for this 5? While I agree we need to settle on a small chunk size to minimize contention, a throughput benchmark would be ideal.

Reply:
Well obviously step = 5 was just to test on a small dataset.
Step is to determine. I'm currently testing on a bigger and bigger dataset (30K -> 600K -> 3M) and I already see a lot of performance improvements (up to 10 times faster for cron execution with 6 to 30 transactions / sec, depending of the number of decimals (0 or 2)). And still improvements to do in the way the probabilities are computed (with 3M records, I get a memory error while doing 
`lead_probabilities leads_to_update._pls_get_naive_bayes_probabilities(batch_mode=True)`)
To be continued..